okhttp/src/jvmMain/kotlin/okhttp3/internal/ws/RealWebSocket.kt

```diff
@@ -304,15 +300,54 @@ class RealWebSocket(
       reader!!.processNextFrame()
       receivedCloseCode == -1
     } catch (e: Exception) {
-      failWebSocket(e, null)
+      failWebSocket(e = e)
       false
     }
   }
 
-  /** For testing: wait until the web socket's executor has terminated. */
-  @Throws(InterruptedException::class)
-  fun awaitTermination(timeout: Long, timeUnit: TimeUnit) {
-    taskQueue.idleLatch().await(timeout, timeUnit)
+  /**
+   * Clean up and publish necessary close events when the reader is done. Invoked only by the reader
+   * thread.
+   */
+  fun finishReader() {
+    val failed: Boolean
+    val code: Int
+    val reason: String?
+    var streamsToClose: Streams?
+    var readerToClose: WebSocketReader?
+    synchronized(this) {
+      failed = this.failed
+      code = receivedCloseCode
+      reason = receivedCloseReason
+
+      readerToClose = reader
+      reader = null
+
+      if (enqueuedClose && messageAndCloseQueue.isEmpty()) {
+        // Close the writer on the writer's thread.
+        val writerToClose = this.writer
+        if (writerToClose != null) {
+          this.writer = null
+          taskQueue.execute("$name writer close", cancelable = false) {
+            writerToClose.closeQuietly()
+          }
+        }
+
+        this.taskQueue.shutdown()
+      }
+
+      streamsToClose = when {
+        writer == null -> streams
+        else -> null
+      }
+    }
+
+    if (!failed && streamsToClose != null && receivedCloseCode != -1) {
+      listener.onClosed(this, code, reason!!)
```

Comment:
This is now published either when the reader exits, or when the writer writes Close, whichever happens last